src/reference_tree.ml

```diff
@@ -0,0 +1,237 @@
+type node_type = Leaf | Tag | Other
+
+type ref_node_data = {
+    node_type: node_type;
+    constraints: (Value_checker.value_constraint list);
+    help: string;
+    value_help: (string * string) list;
+    constraint_error_message: string;
+    multi: bool;
+    valueless: bool;
+    owner: string option;
+    keep_order: bool;
+    hidden: bool;
+    secret: bool;
+}
+
+type t = ref_node_data Vytree.t
+
+exception Bad_interface_definition of string
+
+exception Validation_error of string
+
+let default_data = {
+    node_type = Other;
+    constraints = [];
+    help = "No help available";
+    value_help = [];
+    constraint_error_message = "Invalid value";
+    multi = false;
+    valueless = false;
+    owner = None;
+    keep_order = false;
+    hidden = false;
+    secret = false;
+}
+
+let default = Vytree.make default_data "root"
+
+(* Loading from XML *)
+
+let node_type_of_string s =
+    match s with
+    | "node" -> Other
+    | "tagNode" -> Tag
+    | "leafNode" -> Leaf
+    | _ -> raise (Bad_interface_definition
+                  (Printf.sprintf "node, tagNode, or leafNode expected, %s found" s))
+
+let load_constraint_from_xml d c =
+    let aux d c =
+        match c with
+        | Xml.Element ("regex", _, [Xml.PCData s]) ->
+            let cs = (Value_checker.Regex s) :: d.constraints in
+            {d with constraints=cs}
+        | Xml.Element ("validator", [("name", n); ("argument", a)], _) ->
+            let cs = (Value_checker.External (n, Some a)) :: d.constraints in
+            {d with constraints=cs}
+        | Xml.Element ("validator", [("name", n)], _) ->
+            let cs = (Value_checker.External (n, None)) :: d.constraints in
+            {d with constraints=cs}
+        | _ -> raise (Bad_interface_definition "Malformed constraint")
+    in Xml.fold aux d c
+
+let data_from_xml d x =
+    let aux d x =
+        match x with
+        | Xml.Element ("help", _, [Xml.PCData s]) -> {d with help=s}
+        | Xml.Element ("valueHelp", _,
+                       [Xml.Element ("format", _, [Xml.PCData fmt]);
+                        Xml.Element ("description", _, [Xml.PCData descr])]) ->
+            let vhs = d.value_help in
+            let vhs' = (fmt, descr) :: vhs in
+            {d with value_help=vhs'}
+        | Xml.Element ("multi", _, _) -> {d with multi=true}
+        | Xml.Element ("valueless", _, _) -> {d with valueless=true}
+        | Xml.Element ("constraintErrorMessage", _, [Xml.PCData s]) ->
+            {d with constraint_error_message=s}
+        | Xml.Element ("constraint", _, _) -> load_constraint_from_xml d x
+        | Xml.Element ("hidden", _, _) -> {d with hidden=true}
+        | Xml.Element ("secret", _, _) -> {d with secret=true}
+        | Xml.Element ("keepChildOrder", _, _) -> {d with keep_order=true}
```

Comment:
I wonder if we should have it here while we still use the old backend that always sorts nodes.

Reply:
Dropped.